packages/material-ui/src/Badge/Badge.js

```diff
@@ -1,194 +1,230 @@
+import styled from '../styles/experimentalStyled';
+import useThemeProps from '../styles/useThemeProps';
 import capitalize from '../utils/capitalize';
 
+const badgeClasses = {
+  ...badgeUnstyledClasses,
+  colorError: getBadgeUtilityClass('colorError'),
+  colorPrimary: getBadgeUtilityClass('colorPrimary'),
+  colorSecondary: getBadgeUtilityClass('colorSecondary'),
+};
+
+export { badgeClasses };
+
 const RADIUS_STANDARD = 10;
 const RADIUS_DOT = 4;
 
-export const styles = (theme) => ({
-  /* Styles applied to the root element. */
-  root: {
-    position: 'relative',
-    display: 'inline-flex',
-    // For correct alignment with the text.
-    verticalAlign: 'middle',
-    flexShrink: 0,
-  },
-  /* Styles applied to the badge `span` element. */
-  badge: {
-    display: 'flex',
-    flexDirection: 'row',
-    flexWrap: 'wrap',
-    justifyContent: 'center',
-    alignContent: 'center',
-    alignItems: 'center',
-    position: 'absolute',
-    boxSizing: 'border-box',
-    fontFamily: theme.typography.fontFamily,
-    fontWeight: theme.typography.fontWeightMedium,
-    fontSize: theme.typography.pxToRem(12),
-    minWidth: RADIUS_STANDARD * 2,
-    lineHeight: 1,
-    padding: '0 6px',
-    height: RADIUS_STANDARD * 2,
-    borderRadius: RADIUS_STANDARD,
-    zIndex: 1, // Render the badge on top of potential ripples.
-    transition: theme.transitions.create('transform', {
-      easing: theme.transitions.easing.easeInOut,
-      duration: theme.transitions.duration.enteringScreen,
-    }),
-  },
-  /* Styles applied to the root element if `color="primary"`. */
-  colorPrimary: {
-    backgroundColor: theme.palette.primary.main,
-    color: theme.palette.primary.contrastText,
-  },
-  /* Styles applied to the root element if `color="secondary"`. */
-  colorSecondary: {
-    backgroundColor: theme.palette.secondary.main,
-    color: theme.palette.secondary.contrastText,
-  },
-  /* Styles applied to the root element if `color="error"`. */
-  colorError: {
-    backgroundColor: theme.palette.error.main,
-    color: theme.palette.error.contrastText,
-  },
-  /* Styles applied to the root element if `variant="dot"`. */
-  dot: {
+const overridesResolver = (props, styles) => {
+  const {
+    color = 'default',
+    variant = 'standard',
+    anchorOrigin = {
+      vertical: 'top',
+      horizontal: 'right',
+    },
+    invisible,
+    overlap = 'rectangular',
+  } = props;
+
+  const styleOverrides = {
+    ...styles.root,
+    [`& .${badgeClasses.badge}`]: {
+      ...styles.badge,
+      ...styles[variant],
+      ...styles[
+        `anchorOrigin${capitalize(anchorOrigin.vertical)}${capitalize(
+          anchorOrigin.horizontal,
+        )}${capitalize(overlap)}`
+      ],
+      ...(color !== 'default' && styles[`color${capitalize(color)}`]),
+      ...(invisible && styles.invisible),
+    },
+  };
+
+  return styleOverrides;
+};
+
+const BadgeRoot = styled(
+  'span',
+  {},
+  { muiName: 'MuiBadge', overridesResolver },
```

Comment:
Would this work?

```suggestion
  { muiName: 'MuiBadge-root', overridesResolver },
```

Reply:
This is used as key for the overrides inside the theme, not sure we want to actually change it. Should we maybe add a new prop for the `class` suffix we want to have?

Reply:
I wouldn't be against adding a second option to get this feature. I doubt that developers should be able to override `MuiBadge-badge` with the theme.

Reply:
I've tried to work around this, the problem is that whatever we pass in the `label`, will be used for both the component display name and classname suffix. It would feel awkward seeing component names like `MuiBadge-root`. This is currently on all slots components at the moment. I would even suggest we use as muiName: 'MuiBadge', `MuiSlider`, `MuiSliderTrack` etc, to have better display names on the components. The utility classes are anyway added to the components, so people can easily copy them. 

Let me know what do you think and I'll update the migrated dcomponents.

Reply:
My motivation for the proposal was to mitigate this confusion https://github.com/mui-org/material-ui/pull/23841#pullrequestreview-547156423.

Reply:
Got ya, created https://github.com/mui-org/material-ui/pull/23964 hopefully it will help to resolve the confusion